specs/core/0_beacon-chain.md

```diff
@@ -1771,80 +1774,58 @@ def maybe_reset_eth1_period(state: BeaconState) -> None:
+def get_base_reward(state: BeaconState, total_balance: Gwei, index: ValidatorIndex) -> Gwei:
     if total_balance == 0:
         return 0
 
     adjusted_quotient = integer_squareroot(total_balance) // BASE_REWARD_QUOTIENT
-    return get_effective_balance(state, index) // adjusted_quotient // 5
-```
-
-```python
-def get_base_reward(state: BeaconState, index: ValidatorIndex) -> Gwei:
-    return get_base_reward_from_total_balance(state, get_previous_total_balance(state), index)
-```
-
-```python
-def get_inactivity_penalty(state: BeaconState, index: ValidatorIndex, epochs_since_finality: int) -> Gwei:
-    if epochs_since_finality <= 4:
-        extra_penalty = 0
-    else:
-        extra_penalty = get_effective_balance(state, index) * epochs_since_finality // INACTIVITY_PENALTY_QUOTIENT // 2
-    return get_base_reward(state, index) + extra_penalty
+    return get_effective_balance(state, index) // adjusted_quotient // BASE_REWARDS_PER_EPOCH
 ```
 
-Note: When applying penalties in the following balance recalculations, implementers should make sure the `uint64` does not underflow.
-
 ##### Justification and finalization
 
 ```python
 def get_justification_and_finalization_deltas(state: BeaconState) -> Tuple[List[Gwei], List[Gwei]]:
-    current_epoch = get_current_epoch(state)
-    epochs_since_finality = current_epoch + 1 - state.finalized_epoch
-    rewards = [0 for index in range(len(state.validator_registry))]
-    penalties = [0 for index in range(len(state.validator_registry))]
-    # Some helper variables
-    boundary_attestations = get_previous_epoch_boundary_attestations(state)
-    boundary_attesting_balance = get_attesting_balance(state, boundary_attestations)
-    total_balance = get_previous_total_balance(state)
-    total_attesting_balance = get_attesting_balance(state, state.previous_epoch_attestations)
-    matching_head_attestations = get_previous_epoch_matching_head_attestations(state)
-    matching_head_balance = get_attesting_balance(state, matching_head_attestations)
+    previous_epoch = get_previous_epoch(state)
     eligible_validators = [
         index for index, validator in enumerate(state.validator_registry)
         if (
-            is_active_validator(validator, current_epoch) or
-            (validator.slashed and current_epoch < validator.withdrawable_epoch)
+            is_active_validator(validator, previous_epoch) or
+            (validator.slashed and previous_epoch < validator.withdrawable_epoch)
         )
     ]
-    # Process rewards or penalties for all validators
+    rewards = [0 for index in range(len(state.validator_registry))]
+    penalties = [0 for index in range(len(state.validator_registry))]
     for index in eligible_validators:
-        base_reward = get_base_reward(state, index)
-        # Expected FFG source
+        base_reward = get_base_reward(state, get_previous_epoch_total_balance(state), index)
```

Comment:
So I know this has not changed, but is there an explicit reason to use get_previous_epoch_total_balance rather than the current balance here? This has a nasty edge case in the GENESIS_EPOCH that needs to be fixed:  #941

Reply:
I think I agree with Danny. See [here](https://github.com/ethereum/eth2.0-specs/issues/941#issuecomment-483947322).

Reply:
So why then use the current validator balance, if using the previous total balance? (sorry for the parallel discussion)